brdgrd.c

```diff
@@ -294,13 +305,38 @@ int main( int argc, char **argv ) {
 				break;
 			case 'q': queue_number = atoi(optarg);
 				break;
+			case 'm': window_min = atoi(optarg);
+				break;
+			case 'M': window_max = atoi(optarg);
+				break;
 			case 'h': help(argv[0]);
 				return 0;
 			case '?': help(argv[0]);
 				return 1;
 		}
 	}
 
+	if (window_min < 0 || window_max < 0 || window_min > 0xffff || window_max > 0xffff) {
+		fprintf(stderr, "TCP window is 16bit value.\n");
+		return 1;
+	}
+	if (window_min > window_max) {
+		fprintf(stderr, "Minimum TCP window should be less or equal then maximum one.\n");
+		return 1;
+	}
+	window_rnd_mask = window_max - window_min;
+	if (window_rnd_mask) {
```

Comment:
Why is this necessary? Could you please add a comment explaining the block?

Reply:
For some reason I thought that the following loop
```
	do {
		new_window = 60 + (rand() % 31);
	} while (new_window < 60 || new_window > 90); 
```
had following reason: I thought, that `% 31` here is used instead of `& 0b00011111` and the loop was there to take lowest bits of `rand()` and use them as a random number in specific range to avoid possible skews in distribution where both 0x32 and 0x0 map to 0.

But, now I see that I've likely misunderstood that piece of code as `% 31` is obviously not `& 31` :) Tell me, please, are you worried about small non-uniformity of random numbers when `% non_power_of_two` is used to map `[0, x)` to `[0, non_power_of_two)` ? If not, I'll just place `rand() % range` here.

But, now I don't understand the reason for while() loop to be there... Do you by a chance remember the logic behind it?

Also, the code should double-check in compile time that RAND_MAX can generate full range of uint16_t (that's usually the case, but may be wrong assumption on some systems).

Reply:
The while loop is supposed to find a non-biased random number in the interval [60, 90]. I used 31 as modulo because the expression `rand() % 31` results in a maximum value of 30 and 30 + 60 is 90, which is the maximum window size. I can see how the number 31 is misleading. I should have written it as 90 - 60 + 1. I think you can do the same, and use window_max - window_min + 1. Does that make sense?

Reply:
I'm still a bit confused. As far as I see, `rand() % 31` in old version of code still produced biased random number in the interval.

`RAND_MAX` on my Linux system is 0x7fffffff, so the interval is [0; 0x7fffffff]. So there are 69273666 values having remainders 2, 3, ... 30, while there are 69273667 values having remainders 0 and 1. That's tiny bias, but it's there.

Was it the bias you were trying to avoid? I ask, as I don't see how loop helps to avoid it as `rand() % 31` is always in `[0; 30]` range, so `60 + [0; 30]` is always `[60; 90]`. Am I missing something?

Meanwhile, I'm going to update the patch with more simple code that does not fight for non-biased random as, seems, it's how it actually worked before :)

Reply:
I just had another look and you are right. The while loop is useless. I added it in commit `8e85066` and it hasn't changed since then. I don't remember what I was thinking. Also, I'm not worried about minor biases. I think it's fine to use something like:

    new_window = window_min + (rand() % (window_max - window_min + 1);

What do you think?

Reply:
I also think it's fine. That's the behavior I've updated the patch to.
Moreover, I think it would be "an honor" to see specifically `brdgrd` being distinguished specifically by `rand()` non-uniformity and predictability (if that ever happens) :-)